kroxylicious/src/test/java/io/kroxylicious/proxy/internal/FilterHandlerTest.java

```diff
@@ -17,41 +17,117 @@
+    public void testModifyHeaderRequest() {
+        ApiVersionsRequestFilter filter = (request, context) -> {
+            request.header().setClientId("b");
+            context.forwardRequest(request.body());
+        };
+        buildChannel(KrpcFilter.of(filter));
+        var frame = createRequestFrame(new ApiVersionsRequestData());
+        frame.header().setClientId("a");
+        channel.writeOutbound(frame);
+        var propagated = channel.readOutbound();
+        assertSame(frame, propagated, "Expect it to be the frame that was sent");
+        assertEquals("b", frame.header().clientId(), "Expect clientId header to be mutated by Filter");
+    }
+
+    @Test
+    public void testModifyHeaderAllMessages() {
+        KrpcFilter filter = new KrpcFilter() {
+
+            @Override
+            public void onRequest(DecodedRequestFrame<?> request, KrpcFilterContext context) {
+                request.header().setClientId("b");
+                context.forwardRequest(request.body());
+            }
+
             @Override
             public boolean shouldDeserializeRequest(ApiKeys apiKey, short apiVersion) {
+                return true;
+            }
+
+            @Override
+            public boolean shouldDeserializeResponse(ApiKeys apiKey, short apiVersion) {
                 return false;
             }
 
+        };
+        buildChannel(KrpcFilter.of(filter));
+        var frame = createRequestFrame(new ApiVersionsRequestData());
+        frame.header().setClientId("a");
+        channel.writeOutbound(frame);
+        var propagated = channel.readOutbound();
+        assertSame(frame, propagated, "Expect it to be the frame that was sent");
+        assertEquals("b", frame.header().clientId(), "Expect clientId header to be mutated by Filter");
+    }
+
+    @Test
+    public void testForwardRequestWithKrpcImpl() {
+        KrpcFilter filter = new KrpcFilter() {
+            @Override
+            public boolean shouldDeserializeRequest(ApiKeys apiKey, short apiVersion) {
+                return true;
+            }
+
             @Override
-            public void onApiVersionsRequest(ApiVersionsRequestData request, KrpcFilterContext context) {
-                fail("Should not be called");
+            public boolean shouldDeserializeResponse(ApiKeys apiKey, short apiVersion) {
+                return true;
+            }
+
+            @Override
+            public void onRequest(DecodedRequestFrame<?> decodedFrame, KrpcFilterContext filterContext) {
+                filterContext.forwardRequest(decodedFrame.body());
             }
         };
-        buildChannel(filter);
+        buildChannel(KrpcFilter.of(filter));
+        var frame = writeRequest(new ApiVersionsRequestData());
+        var propagated = channel.readOutbound();
+        assertEquals(frame, propagated, "Expect it to be the frame that was sent");
+    }
+
+    private static class NoDeserialiseFilter implements ApiVersionsRequestFilter, RequestTargeter {
```

Comment:
maybe instead of having separate interfaces for targeting all the generated Filter interfaces should have a method like: `shouldDeserializeApiVersionsRequest(short apiVersion)` defaulted to true. Then you could override the method for each type rather than having to switch on `apiKey`

Reply:
I think I prefer this idea to the Targeters.  Alternatively someone expressed an idea about annotating the method with something that conveys the same information.